### PR TITLE
Check index vector length in partialsortperm!(); clarify documentation

### LINKED ISSUE #33184: partialsortperm! silently gives wrong result
I believe I have found a silent bug with `partialsortperm!` (almost went nuts debugging the huge code base where this was happening). Here is the MRE (compare the outputs of the `partialsortperm!` with `partialsortperm` and the outputs of `v[ix]` in both cases):
```Julia
julia> v = [3, 1, 5, 2, 4,
            3, 1, 5, 2, 4]
10-element Array{Int64,1}:
 3
 1
 5
 2
 4
 3
 1
 5
 2
 4

julia> k = 1:5
1:5

julia> ix = collect(k)
5-element Array{Int64,1}:
 1
 2
 3
 4
 5

julia> partialsortperm!(ix, v, k)
5-element view(::Array{Int64,1}, 1:5) with eltype Int64:
 2
 4
 1
 5
 3

julia> v[ix]
5-element Array{Int64,1}:
 1
 2
 3
 4
 5

julia> ix = partialsortperm(v, k)
5-element view(::Array{Int64,1}, 1:5) with eltype Int64:
 2
 7
 4
 9
 1

julia> v[ix]
5-element Array{Int64,1}:
 1
 1
 2
 2
 3
```
```Julia
julia> versioninfo()
Julia Version 1.2.0
Commit c6da87ff4b (2019-08-20 00:03 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.6.0)
  CPU: Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, haswell)
Environment:
  JULIA_EDITOR = atom  -a
  JULIA_NUM_THREADS = 4
```

### PR DESCRIPTION
**Pull request by kmsquire**
* Check that index vector has the same axes as the source vector
* Clarify documentation and add doc-tests

Fixes #33184

### CODE REVIEW

File: base/sort.jl [hunk 1]
```diff
@@ -789,16 +789,51 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
-"""
+Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix` the same size as
+`v`.
+
+If the index vector `ix` is initialized with the indices of `v`, `initialized` should be set to
+`true`.
+
+If `initialized` is `false` (the default), then `ix` is initialized to contain the indices of `v`.
+
+(Typically, the indices of `v` will be `1:length(ix)`, although if `v` is an alternative array types
```
Comment (dkarrasch) [inline base/sort.jl]: "is of an alternative array type"? (add "of", remove "s")

Comment (kmsquire) [reply base/sort.jl]:
```suggestion
(Typically, the indices of `v` will be `1:length(ix)`, although if `v` is an alternative array type
```

Comment (dkarrasch) [reply base/sort.jl]: I'm not a native speaker, but I thought that `v` is an array, not an array type. So, isn't an array "of some (array) type"?

Comment (kmsquire) [reply base/sort.jl]: Sorry, I was in too much of a hurry.  An array can be "of a type", as you suggested, or we can say it "has a type", which I think works better here.

File: base/sort.jl [hunk 2]
```diff
@@ -789,16 +789,51 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
-"""
+Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix` the same size as
+`v`.
+
+If the index vector `ix` is initialized with the indices of `v`, `initialized` should be set to
+`true`.
+
+If `initialized` is `false` (the default), then `ix` is initialized to contain the indices of `v`.
+
+(Typically, the indices of `v` will be `1:length(ix)`, although if `v` is an alternative array type
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
(Typically, the indices of `v` will be `1:length(ix)`, although if `v` has an alternative array type
```

**Comment by haampie** [comment]
How about calling 'index vector of the size of the vector v' a 'permutation of the indices of v'?

File: base/sort.jl (hunk 1 above)
Comment (haampie) [reply base/sort.jl]: Just maybe `1:length(ix)` could be replaced by `1:length(v)`, just to highlight that the size of `ix` depends on the size of `v`?

File: base/sort.jl [hunk 3]
```diff
@@ -789,16 +789,53 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+# Examples
+```julia-repl
+julia> v = [3, 1, 2, 1];
+
+julia> ix = Vector{Int}(undef, 4);
+
+julia> partialsortperm!(ix, v, 1)
+2
+
+julia> ix = [1:4...];
```
Comment (JeffBezanson) [inline base/sort.jl]:
```suggestion
julia> ix = [1:4;];
```

File: base/sort.jl [hunk 4]
```diff
@@ -789,16 +789,53 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+`v`.
+
+If the index vector `ix` is initialized with the indices of `v`, `initialized` should be set to
+`true`.
+
+If `initialized` is `false` (the default), then `ix` is initialized to contain the indices of `v`.
+
+If `initialized` is `true`, but `ix` does not contain the indices of `v`, the behavior is undefined.
+
+(Typically, the indices of `v` will be `1:length(ix)`, although if `v` has an alternative array type
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
(Typically, the indices of `v` will be `1:length(v)`, although if `v` has an alternative array type
```

File: base/sort.jl [hunk 5]
```diff
@@ -789,16 +789,53 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+
+If the index vector `ix` is initialized with the indices of `v`, `initialized` should be set to
+`true`.
+
+If `initialized` is `false` (the default), then `ix` is initialized to contain the indices of `v`.
+
+If `initialized` is `true`, but `ix` does not contain the indices of `v`, the behavior is undefined.
+
+(Typically, the indices of `v` will be `1:length(v)`, although if `v` has an alternative array type
+with non-one-based indices, such as an `OffsetArray`, `ix` must have the same indices.)
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
with non-one-based indices, such as an `OffsetArray`, `ix` must also be an `OffsetArray` with the same
indices, and the contents of `ix` must be these same indices, or a permutation thereof.)
```

File: base/sort.jl [hunk 6]
```diff
@@ -789,16 +789,53 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
 """
     partialsortperm!(ix, v, k; by=<transform>, lt=<comparison>, rev=false, initialized=false)
 
-Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix`. If `initialized` is `false`
-(the default), `ix` is initialized to contain the values `1:length(ix)`.
-"""
+Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix` the same size as
+`v`.
+
+If the index vector `ix` is initialized with the indices of `v`, `initialized` should be set to
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
If the index vector `ix` is initialized with the indices of `v` (or a permutation thereof), `initialized` should be set to
```

File: base/sort.jl [hunk 7]
```diff
@@ -789,16 +789,53 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
 """
     partialsortperm!(ix, v, k; by=<transform>, lt=<comparison>, rev=false, initialized=false)
 
-Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix`. If `initialized` is `false`
-(the default), `ix` is initialized to contain the values `1:length(ix)`.
-"""
+Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix` the same size as
+`v`.
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
`v`, which contains (a permutation of) the indices of `v`. 
```

File: base/sort.jl [hunk 8]
```diff
@@ -789,16 +789,53 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+
+If the index vector `ix` is initialized with the indices of `v`, `initialized` should be set to
+`true`.
+
+If `initialized` is `false` (the default), then `ix` is initialized to contain the indices of `v`.
+
+If `initialized` is `true`, but `ix` does not contain the indices of `v`, the behavior is undefined.
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
If `initialized` is `true`, but `ix` does not contain a permutation of the indices of `v`, the behavior of
`partialsortperm!` is undefined.
```

File: base/sort.jl [hunk 9]
```diff
@@ -789,16 +789,55 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+`true`.
+
+If `initialized` is `false` (the default), then `ix` is initialized to contain the indices of `v`.
+
+If `initialized` is `true`, but `ix` does not contain (a permutation of) the indices of `v`, the behavior of
+`partialsortperm!` is undefined.
+
+(Typically, the indices of `v` will be `1:length(v)`, although if `v` has an alternative array type
+with non-one-based indices, such as an `OffsetArray`, `ix` must also be an `OffsetArray` with the same
+indices, and the contents of `ix` must be these same indices, or a permutation thereof.)
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
indices, and must contain as values (a permutation of) these same indices.)
```

File: base/sort.jl [hunk 10]
```diff
@@ -789,16 +789,55 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+
+Upon return, `ix` is guaranteed to have the indices `k` in their sorted positions, such that
+`v[partialsortperm!(ix, v, k)] == partialsort!(v, k)`
+
+The return value is the `k`th element of `ix` if `k` is an Integer, or view into `ix` if `k` is
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
The return value is the `k`th element of `ix` if `k` is an integer, or view into `ix` if `k` is
```

File: base/sort.jl [hunk 11]
```diff
@@ -789,16 +789,55 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
 """
     partialsortperm!(ix, v, k; by=<transform>, lt=<comparison>, rev=false, initialized=false)
 
-Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix`. If `initialized` is `false`
-(the default), `ix` is initialized to contain the values `1:length(ix)`.
-"""
+Like [`partialsortperm`](@ref), but accepts a preallocated index vector `ix` the same size as
+`v`, which is used to store (a permutation of) the indices of `v`.
```
Comment (kmsquire) [inline base/sort.jl]:
```suggestion
`v`, which is used to store (a permutation of) the indices of `v`.
```

File: base/sort.jl [hunk 12]
```diff
@@ -789,16 +789,55 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+julia> partialsortperm!(ix, v, 1)
+2
+
+julia> ix = [1:4;];
+
+julia> partialsortperm!(ix, v, 2:3, initialized=true)
+2-element view(::Array{Int64,1}, 2:3) with eltype Int64:
+ 4
+ 3
+
```
Comment (dkarrasch) [inline base/sort.jl]: Should there be triple backticks here to close the code block?

File: base/sort.jl [hunk 13]
```diff
@@ -789,16 +789,55 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+indices, and must contain as values (a permutation of) these same indices.)
+
+Upon return, `ix` is guaranteed to have the indices `k` in their sorted positions, such that
+`v[partialsortperm!(ix, v, k)] == partialsort!(v, k)`
+
+The return value is the `k`th element of `ix` if `k` is an integer, or view into `ix` if `k` is
+a range.
+
+# Examples
+```julia-repl
```
Comment (fredrikekre) [inline base/sort.jl]:
```suggestion
```jldoctest
```

File: base/sort.jl [hunk 14]
```diff
@@ -789,16 +789,55 @@ partialsortperm(v::AbstractVector, k::Union{Integer,OrdinalRange}; kwargs...) =
...
+If `initialized` is `false` (the default), then `ix` is initialized to contain the indices of `v`.
+
+If `initialized` is `true`, but `ix` does not contain (a permutation of) the indices of `v`, the behavior of
+`partialsortperm!` is undefined.
+
+(Typically, the indices of `v` will be `1:length(v)`, although if `v` has an alternative array type
+with non-one-based indices, such as an `OffsetArray`, `ix` must also be an `OffsetArray` with the same
+indices, and must contain as values (a permutation of) these same indices.)
+
+Upon return, `ix` is guaranteed to have the indices `k` in their sorted positions, such that
```
Comment (KristofferC) [inline base/sort.jl]: Would 

```jl
partialsortperm!(ix, v, k)
v[ix[k]] == partialsort!(v, k)
```

be more clear here?

Comment (KristofferC) [reply base/sort.jl]: My point was also that the example is talking about guarantees of the state of `ix` on return but then doesn't use that directly, but instead uses the return value which is not identical.

**Comment by kmsquire** [comment]
Good to merge?

**Review by dkarrasch — APPROVED** [review]
LGTM

**Pull request merged**